### PR TITLE
Improve attack model types in inference attacks

### LINKED ISSUE #2153: Support for additional attack model types in inference attacks
Blackbox membership inference attack currently supports 3 model types: nn, rf and gb.
Blackbox attribute inference attack currently supports 2 model types: nn and rf.
Both could benefit from supporting additional attack model types (e.g., KNN, LR, etc.)

### PR DESCRIPTION
**Pull request by abigailgold**
# Description

Support additional attack model types (e.g., KNN, LR, etc.) in both membership inference and attribute inference attacks (blackbox and baseline).
Replace use of sklearn MLPClassifier in attribute attacks with pytorch NN model to improve performance.

Fixes #2153 

## Type of change

Please check all relevant options.

- [X] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

Added to existing tests the new model params. All tests pass.

**Test Configuration**:
- OS: MacOS 12.6.8
- Python version: 3.9

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 1]
```diff
@@ -279,25 +366,92 @@ def infer(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> np.n
...
+            from torch.utils.data import DataLoader
+            from art.utils import to_cuda, from_cuda
+
+            self.attack_model.eval()  # type: ignore
+            predictions: np.ndarray = np.array([])
+            test_set = self._get_attack_dataset(feature=x_test)
+            test_loader = DataLoader(test_set, batch_size=self.batch_size, shuffle=False, num_workers=0)
+            for input1, _ in test_loader:
+                input1 = to_cuda(input1)
+                outputs = self.attack_model(input1)  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/true_label_baseline.py:379]: Why is this type ignore needed?

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 2]
```diff
@@ -240,7 +229,105 @@ def fit(self, x: np.ndarray, y: np.ndarray) -> None:
...
+            self.attack_model = to_cuda(self.attack_model)  # type: ignore
+            self.attack_model.train()  # type: ignore
+
+            for _ in range(self.epochs):
+                for (input1, targets) in train_loader:
+                    input1, targets = to_cuda(input1), to_cuda(targets)
+                    _, targets = torch.autograd.Variable(input1), torch.autograd.Variable(targets)
+
+                    optimizer.zero_grad()
+                    outputs = self.attack_model(input1)  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/true_label_baseline.py:325]: Why is this type ignore needed?

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 3]
```diff
@@ -240,7 +229,105 @@ def fit(self, x: np.ndarray, y: np.ndarray) -> None:
...
+
+                self.attack_model = MembershipInferenceAttackModel(x_train.shape[1], len(self._values))
+                loss_fn = nn.CrossEntropyLoss()
+
+            optimizer = optim.Adam(self.attack_model.parameters(), lr=self.learning_rate)  # type: ignore
+
+            attack_train_set = self._get_attack_dataset(feature=x_train, label=y_ready)
+            train_loader = DataLoader(attack_train_set, batch_size=self.batch_size, shuffle=True, num_workers=0)
+
+            self.attack_model = to_cuda(self.attack_model)  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/true_label_baseline.py:316]: Why is this type ignore needed?

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 4]
```diff
@@ -240,7 +229,105 @@ def fit(self, x: np.ndarray, y: np.ndarray) -> None:
...
+
+                    def forward(self, x):
+                        """Forward the model."""
+                        out = self.features(x)
+                        return self.output(out)
+
+                self.attack_model = MembershipInferenceAttackModel(x_train.shape[1], len(self._values))
+                loss_fn = nn.CrossEntropyLoss()
+
+            optimizer = optim.Adam(self.attack_model.parameters(), lr=self.learning_rate)  # type: ignore
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/true_label_baseline.py:311]: Why is this type ignore needed?

File: art/attacks/inference/attribute_inference/baseline.py [hunk 5]
```diff
@@ -90,10 +100,15 @@ def __init__(
         """
         super().__init__(estimator=None, attack_feature=attack_feature)
 
-        self._values: Optional[list] = None
+        self._values: list = []
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/baseline.py:103]: Empty lists should not be default values for arguments because if multiple instances of this class are created they will overwrite each others arguments.
```suggestion
        self._values: Optional[list] = None
```

File: art/attacks/inference/attribute_inference/black_box.py [hunk 6]
```diff
@@ -106,12 +116,19 @@ def __init__(
         """
         super().__init__(estimator=estimator, attack_feature=attack_feature)
 
-        self._values: Optional[list] = None
-        self._attack_model_type = attack_model_type
-        self._attack_model = attack_model
+        self._values: list = []
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/black_box.py:119]: Empty lists should not be default values for arguments because if multiple instances of this class are created they will overwrite each others arguments.
```suggestion
        self._values: Optional[list] = None
```

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 7]
```diff
@@ -102,10 +112,15 @@ def __init__(
         """
         super().__init__(estimator=None, attack_feature=attack_feature)
 
-        self._values: Optional[list] = None
+        self._values: list = []
```
[the remark this answers is not in the document]
Comment (abigailgold) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:115]: This is not an argument but an assignment within the method, I don't think the same issue exists. When using an optional value it caused mypy errors.

File: art/attacks/inference/attribute_inference/true_label_baseline.py (hunk 2 above)
Comment (abigailgold) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:325]: It was there before (in other classes), I just copied the code. Not sure why it's needed (I can tru to remove and see if it causes any errors).

File: tests/attacks/inference/attribute_inference/test_black_box.py [hunk 8]
```diff
@@ -124,15 +124,21 @@ def test_black_box_continuous(art_warning, decision_tree_estimator, get_iris_dat
         inferred_train = attack.infer(x_train_for_attack, pred=x_train_predictions)
         inferred_test = attack.infer(x_test_for_attack, pred=x_test_predictions)
         # check accuracy
-        assert np.allclose(inferred_train, x_train_feature.reshape(1, -1), atol=0.4)
-        assert np.allclose(inferred_test, x_test_feature.reshape(1, -1), atol=0.4)
+        assert (
+            np.count_nonzero(np.isclose(inferred_train, x_train_feature.reshape(1, -1), atol=0.4))
+            > inferred_train.shape[0] * 0.75
```
[the remark this answers is not in the document]
Comment (abigailgold) [reply tests/attacks/inference/attribute_inference/test_black_box.py:129]: This is just a relatively relaxed way of measuring accuracy for a regression task. Since not 100% of the values are inferred correctly, we check that at least 75% are (within a given distance).

File: tests/attacks/inference/attribute_inference/test_true_label_baseline.py [hunk 9]
```diff
@@ -308,8 +314,8 @@ def test_true_label_baseline_regression(art_warning, get_diabetes_dataset, model
             baseline_inferred_test
         )
 
-        assert 0.6 <= baseline_train_acc
-        assert 0.6 <= baseline_test_acc
+        assert 0.45 <= baseline_train_acc
```
[the remark this answers is not in the document]
Comment (abigailgold) [reply tests/attacks/inference/attribute_inference/test_true_label_baseline.py:317]: Yes, I think this was the reason for reducing the value. How do you suggest to deal with it?

File: tests/attacks/inference/membership_inference/test_black_box.py [hunk 10]
```diff
@@ -38,6 +38,7 @@
 num_classes_mnist = 10
 
 
+@pytest.mark.skip_framework("scikitlearn")
```
[the remark this answers is not in the document]
Comment (abigailgold) [reply tests/attacks/inference/membership_inference/test_black_box.py:41]: You're right, I will remove it.

File: art/attacks/inference/attribute_inference/true_label_baseline.py (hunk 2 above)
Comment (abigailgold) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:325]: Removing it causes a mypy error: "Tensor" not callable

File: tests/attacks/inference/attribute_inference/test_true_label_baseline.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply tests/attacks/inference/attribute_inference/test_true_label_baseline.py:317]: I was thinking of it could be worth having a dictionary of expectation value like
```python
expected_accuracy = {"nn": 0.65, "rf": 0.45, ...}
```
and use it like
```python
assert expected_accuracy[model_type] == baseline_train_acc 
```

File: art/attacks/inference/attribute_inference/true_label_baseline.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:115]: You are right, I was looking at the wrong line. Please ignore.

File: art/attacks/inference/attribute_inference/true_label_baseline.py [hunk 11]
```diff
@@ -102,10 +112,15 @@ def __init__(
...
         self._encoder = encoder
         self._non_numerical_features = non_numerical_features
         self._is_continuous = is_continuous
+        self._attack_model_type: Optional[str] = attack_model_type
+        self.attack_model: Optional[Any] = None
+        self.epochs = 100
+        self.batch_size = 100
+        self.learning_rate = 0.0001
```
[the remark this answers is not in the document]
Comment (beat-buesser) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:123]: Do you often adjust the parameters to optimise the attack?

File: art/attacks/inference/attribute_inference/true_label_baseline.py (hunk 2 above)
Comment (beat-buesser) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:325]: Ok, leave it for now.

File: art/attacks/inference/attribute_inference/black_box.py (hunk 6 above)
Comment (beat-buesser) [reply art/attacks/inference/attribute_inference/black_box.py:119]: Please ignore.

File: art/attacks/inference/attribute_inference/baseline.py (hunk 5 above)
Comment (beat-buesser) [reply art/attacks/inference/attribute_inference/baseline.py:103]: Please ignore.

File: art/attacks/inference/attribute_inference/true_label_baseline.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (abigailgold) [reply art/attacks/inference/attribute_inference/true_label_baseline.py:123]: We don't, but perhaps a more savvy user would want to... I can do this with default values and then there is no harm... People can either use it or not.

**Review by beat-buesser — APPROVED** [review]
